### ISSUE TITLE
A question about [m/s/vs]status.[sd,fs]

### ISSUE THREAD

**Issue by tianhu-zeng** [riscv-software-src/riscv-isa-sim#553]
Hi, 
    When I verify our hypersior extentation by Spike. There are confusions about the FS/SD fields and vsstatus. 
    1. From Spec description, the SD bit is a read-only bit, but it's writable in Spike
    2. In Spike, there is not the SD bit updating logic. Does it update same as the mstatus.SD?
    3. In VS mode, access "sstatus" actually access "vsstatus" instead. It seems that Spike access sstatus directly.
    4. In Spec, "In systems that do not implement S-mode and do not have a floating-point unit, the FS field is
hardwired to zero". I'm curious, in system with M/S/U mode and without F-extentation, what is FS's function? Why don't need to check F-extentation supproted when update [M/S]STATUS.FS. But vsstatus will check F-extentation supproted.

![image](https://user-images.githubusercontent.com/59038826/94353901-9ccb3180-00a8-11eb-82f9-5c26c9a37246.png)
![image](https://user-images.githubusercontent.com/59038826/94353924-d9972880-00a8-11eb-8f77-0e266b27a7d2.png)
![image](https://user-images.githubusercontent.com/59038826/94353857-51b11e80-00a8-11eb-9a6f-8df0502b038b.png)
![image](https://user-images.githubusercontent.com/59038826/94354247-f123e080-00ab-11eb-9687-f8a2ce7ad663.png)

**Comment by aswaterman** [comment]
I'm just writing to note that the SD logic appears correct for mstatus.  I'll leave the rest to @avpatel.

**Comment by tianhu-zeng** [comment]
Another issue for hypervisor. when mcounteren = 1 and hcounteren = 0,  access user Counter/Timers csrs should report virtual instruction exception in VS/VU. It seems that the logic is not complete in Spike.

**Comment by aswaterman** [comment]
@tianhu-zeng Please don’t create additional confusion by reporting multiple issues together. File a second issue if it isn’t related to the first.

**Comment by tianhu-zeng** [comment]
ok, I'll pay more attention to them next time.

**Comment by avpatel** [comment]
For 1) and 2), I think @aswaterman already answered your question.

For 3), Spike uses CSR swapping approach for emulating MSTATUS, SSTATUS and VSSTATUS. This is done due to complicated nature of all xSTATUS CSRs. When V=0, the S-mode bits in MSTATUS represent HS-mode SSTATUS and S-mode bits in VSSTATUS represent VS-mode SSTATUS. When V=1, the S-mode bits in MSTATUS represent VS-mode SSTATUS and S-mode bits in VSSTATUS represetn HS-mode SSTATUS. When transitioning from V=0 to V=1 (or vice-versa), we swap the S-mode bits between MSTATUS and VSTATUS. This way the MSTATUS and SSTATUS emulation in Spike remain same for both HS-mode and VS-mode.

For 4), The MSTATUS and SSTATUS emulation ensures FS bits are writeable only when 'F' or 'D' extension is available (Refer, MSTATUS and SSTATUS emulation in set_csr() function). The SSTATUS can be viewed as restricted view of MSTATUS that's why set_csr() recursively calls itself for emulating SSTATUS. On other hand, VSSTATUS is a separate CSR so set_csr() emulates it separately and also checks for 'F' and 'D' extension availability.

I hope above addresses your queries. Let me know if I missed anything.

For questions or issues not related to this topic, please create another GitHub issue.

Regards,
Anup

**Comment by tianhu-zeng** [comment]
@avpatel thanks for your reply. I have no doubt about mstatus.sd logic. Maybe I didn't describe my problem clearly.
1. SD bit is a RO field, but it's writable on vsstatus in  Spike. I think it's not make sense, right?
2. There are different conditions about writing FS field on mstatus/status/vstatus in Spike.
     1) mstatus: has_fs = supports_extentation('S') || supports_extentation('F') || supports_extentation('V').
     2) sstatus: can write directly.
     3) vsstatus: supports_extentation('F').
![image](https://user-images.githubusercontent.com/59038826/94380668-10347800-0169-11eb-977a-b292cc7a7ba0.png)
     Different conditions about writing FS field on mstatus/status/vstatus, does it make sense?
3. The following is my understanding about executing "csrw sstatus, x5" in VS mode.
     After executing this instruction, x5 is written to 'vsstatus' and shouldn't affect mstatus/sstatus. I write a case that execute "csrw sstatus, t1" in VS mode. I found that this writing affect 'mstatus/status' csrs, such as 'SD' bit. Following is the testcase and spike.log.
![image](https://user-images.githubusercontent.com/59038826/94386282-d3bd4800-0179-11eb-8f73-54e3e2a5e2ed.png)

Regards,
Tianhu

**Comment by avpatel** [comment]
1. Yes, SD bit of VSSTATUS should be treated as read-only in set_csr() implementation of processor.cc. This needs to be fixed.
2. The condition for writeability of MSTATUS.FS field in MSTATUS emulation is not updated as part of H-extension support in Spike (@aswaterman ??). The SSTATUS.FS and VSSTATUS.FS condition for writeabiltiy looks correct to me.
3. Your understanding is correct but we have taken CSR-swapping approach for MSTATUS/SSTATUS/VSSTATUS emulation. This approach provides functionally same behavior which you described. As-per this approach, when V=1 the SSTATUS represents VS-mode SSTATUS and HS-mode SSTATUS is saved in VSSTATUS CSR whereas when V=0 the SSTATUS represents HS-mode SSTATUS and VS-mode SSTATUS is saved in VSSTATUS CSR.

Regards,
Anup

**Comment by aswaterman** [comment]
@avpatel I don't think the H-extension should affect the writability of mstatus.FS.  The condition for mstatus.FS' writability looks correct to me.

**Comment by tianhu-zeng** [comment]
@avpatel @aswaterman thanks for your reply very much, it's very helpful to me. I'm very curious why [MSTATUS/SSTATUS/VSSTATUS].FS have different writability.

**Comment by avpatel** [comment]
@aswaterman we have not modified mstatus.FS writability at time of adding H-extension to Spike. I guess we are good here. Can you check the writeablitiy condition for VSSTATUS.FS ?

**Comment by aswaterman** [comment]
I see three problems:
- `set_virt` is treating the field like it's Boolean by doing things like `if ((state.mstatus & SSTATUS_FS) == SSTATUS_FS) state.vsstatus |= SSTATUS_FS;`.  This has the effect of turning the Initial and Clean states into Dirty.  Instead, the two-bit contents of the field should be copied, perhaps using the `get_field` macro.
- For the same reason, `set_virt` is setting SD=1 for FS=Initial or Clean, when in fact it should only do so for Dirty.
- It looks like the hypervisor is allowed to write the vsstatus.SD field.  Instead, on a vsstatus write, SD should be set to 1 iff FS/FS/XS is Dirty.  See, for comparison, how mstatus.SD is set: https://github.com/riscv/riscv-isa-sim/blob/f1c24eff543e6f41980993f41ae1ab5ab80a7340/riscv/processor.cc#L872

**Comment by tianhu-zeng** [comment]
@aswaterman @avpatel So, how should we deal with this issue?

**Comment by avpatel** [comment]
I have submitted fix this https://github.com/riscv/riscv-isa-sim/pull/568

Regards,
Anup

**Comment by tianhu-zeng** [comment]
@avpatel I have two questions. 
1. When are you going to push the fixing to the master? 
2. Why need check supports_extension("F") for vsstatus and didn't for sstatus?

**Comment by aswaterman** [comment]
I think checking supports_extension("F") is actually not necessary because of the clause in the spec that FS exists if S-mode exists, even if the F extension doesn't.

**Comment by tianhu-zeng** [comment]
@aswaterman Thanks for your reply. If so, Spike implementation match with Spec. Thanks again.

**Comment by tianhu-zeng** [comment]
@avpatel When could you push the fixing version?

**Comment by aswaterman** [comment]
Anup already made a PR. It is waiting on my review. Maybe in 2-3 days.

**Comment by tianhu-zeng** [comment]
Got it. Does The Spike boot kvm or xvisor successfully?

**Comment by avpatel** [comment]
Yes, both KVM and Xvisor work fine with the mentioned GitHub PR.

Regards,
Anup

**Comment by tianhu-zeng** [comment]
Has this issue been fixed?

**Comment by aswaterman** [comment]
Yes, it was fixed in #568.
